File: ts/src/handlers/http/allow_methods.ts
import { ApiHandler, ApiHandlerUnTyped } from "../../types/app";
import { HttpStatusCode } from "../../types";

export function AllowMethods(...methods: string[]): ApiHandlerUnTyped {
  return function (api: ApiHandler): Promise<boolean> {
    if (methods.length === 0) {
      return Promise.resolve(true);
    }
    const r = api.Request();
    for (const method of methods) {
      if (r.Method === method) {
        return Promise.resolve(true);
      }
    }
    const w = api.Response();
    w.WriteHeader(HttpStatusCode.METHOD_NOT_ALLOWED);
    return Promise.resolve(false);
  };
}

File: go/handlers/http/allow_methods.go
package handlers_http

import types_app "github.com/mabels/runified/types/app"

func AllowMethods(methods ...string) types_app.ApiHandlerUnTyped {
	return func(api types_app.ApiHandler) bool {
		if len(methods) == 0 {
			return true
		}
		r := api.Request()
		for _, method := range methods {
			if r.Method == method {
				return true
			}
		}
		w := api.Response()
		w.WriteHeader(405)
		return false
	}
}
